[android_intent] Ignoring packages in intent.

NikulinEB opened the pull request:
## Description

This update allows to specify a list of packages that should not be used to resolve the intent. For example, this is useful with links to the Steam store. When click on such a link, the Steam application appears in the default selection dialog, but the application crashes when opening a link. This happens due to the lack of a component name in the intent. Therefore, it would be more convenient to exclude the Steam app from this list to avoid crashes when opening a link. And offer the user a choice: open in the Steam app (using the intent with the specified package and component names) or open in a browser (by default or from the list of available browsers without the Steam app). And in this update, a parameter is added to force the default selection dialog to be displayed, if necessary.

## Related Issues

I did not find or create an issue about this. But in the previous paragraph I described the situation in which this update is useful.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -106,6 +111,9 @@ void setApplicationContext(@Nullable Context applicationContext) {
    * @param packageName forwarded to {@link Intent#setPackage(String)} if non-null. This is forced
    *     to null if it can't be resolved.
    * @param componentName forwarded to {@link Intent#setComponent(ComponentName)} if non-null.
+   * @param ignoredPackages list of package names that should not by displayed in the chooser and
```
type: "by" -> "be"

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -106,6 +111,9 @@ void setApplicationContext(@Nullable Context applicationContext) {
    * @param componentName forwarded to {@link Intent#setComponent(ComponentName)} if non-null.
+   * @param ignoredPackages list of package names that should not by displayed in the chooser and
+   *     should not be used to resolve this intent.
+   * @param showChooser forces to show the default selection dialog.
```
What happens if there's no UI/`Activity` available and `showChooser` is set to `true`?

NikulinEB replied:
If there is no activity available, then `targetIntents` will be empty and the сhooser will not be created. The intent will be created as if `showChooser` would be `false`.

matthew-carroll replied:
Can you add that information to the docs?

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,6 +169,45 @@ Intent buildIntent(
         intent.setPackage(null);
       }
     }
+    Boolean needIgnore = ignoredPackages != null && ignoredPackages.size() > 0;
```
Can we clean up this code a bit for readability? It's difficult to quickly see the relevant behavior vs the low level execution. Based on what I see, this could probably be reduced to something like:

```
boolean isPackageFilterRequired = ignoredPackages != null && ignoredPackages.size() > 0;

if (isPackageFilterRequired || isChooserDesired) {
  List<Intent> targetIntents = calculateTargetIntents(...);
  
  if (isChooserDesired) {
    String choosePackageToLaunch(targetIntents);
    intent.setPackage(packageToLaunch);
  } else {
    //...The code that would go here has other issues that I'll comment on
    // but this example should give you an idea of the functional decomposition
    // that I'm thinking of.
  }
}

return intent;
```

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,6 +169,45 @@ Intent buildIntent(
+          Log.v(TAG, "Not found packages to resolve this intent: " + exception);
+        }
+      } else {
+        intent = Intent.createChooser(targetIntents.remove(0), "");
```
I have 2 concerns on this line. First, I don't thin we should be replacing the `intent` value. It is very easy for someone to miss this line and believe that `intent` always refers to the same instance throughout this method.

Second, is there a reason that `remove(0)` is being called? If not, please consider changing it to `get(0)`. If there is a reason, then please consider moving that call to its own line and commenting as to why we need `remove()` instead of `get()`.

NikulinEB replied:
I call `remove(0)` because I then call the `putExtra` method with the entire array of packages. If I do not remove the first intent, then the corresponding package will be in the chooser twice. I could call `createChooser` with `new Intent()` as the first argument, but I'm not sure that is good.

mklim replied:
It also took me a bit of reading to figure out why we were popping the head from the list here and if it was a mistake. I think the decision makes sense, but an implementation comment would be helpful.

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,6 +169,45 @@ Intent buildIntent(
+          }
+          intent.setPackage(packageToLaunch);
+          Log.v(TAG, "Resolve this intent with package: " + packageToLaunch);
+        } catch (IndexOutOfBoundsException exception) {
```
I think I would suggest checking the size explicitly before accessing an item and then voiding the try/catch altogether.

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,6 +169,45 @@ Intent buildIntent(
+          intent.setPackage(packageToLaunch);
+          Log.v(TAG, "Resolve this intent with package: " + packageToLaunch);
+        } catch (IndexOutOfBoundsException exception) {
+          Log.v(TAG, "Not found packages to resolve this intent: " + exception);
```
This log should either be removed or improved.

If this log refers to an expectation that the developer has, and that expectation is not met, then I think this should probably be a "warning" at minimum. 

Also the `IndexOutOfBoundsException` is unlikely to help a developer. Instead, this log could say something like:
`"Intent has no available targets. Some targets were explicitly ignored: " + Arrays.toString(ignoredPackages)`

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,6 +169,45 @@ Intent buildIntent(
+            packageToLaunch = activities.get(0).activityInfo.packageName;
+          }
+          intent.setPackage(packageToLaunch);
+          Log.v(TAG, "Resolve this intent with package: " + packageToLaunch);
```
Please remove this log.

NikulinEB commented:
Thanks for the comments! I updated the pull request.

ened commented:
@NikulinEB @matthew-carroll the launchIntent method seems already a bit complex. Do you think it's more approachable to have a new "showChooser" method which handles the appropriate use case only?

NikulinEB commented:
@ened Good suggestion, thanks!
I added the `showChooser` method, several tests, and updated the pull request.

NikulinEB commented:
Hi @matthew-carroll @mklim 
Could you review this PR, please?
Thanks in advance!

mklim commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,7 +185,46 @@ Intent buildIntent(
         intent.setPackage(null);
       }
     }
+    Boolean isPackageFilterRequired = ignoredPackages != null && ignoredPackages.size() > 0;
```
nit: the primitive `boolean` would be preferable here, confusingly `Boolean` is an object and could theoretically be null.

mklim commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,7 +185,46 @@ Intent buildIntent(
     return intent;
   }
+
+  String choosePackageToLaunch(Intent intent, List<String> ignoredPackages) {
```
This and the below helper methods should be `private`, and I think also `static`?

NikulinEB replied:
I made them only private, but not static, because they use the instance fields `activity` and `applicationContext`, which I would have to pass through several static methods.

mklim commented on packages/android_intent/lib/android_intent.dart:
```diff
@@ -27,6 +27,8 @@ class AndroidIntent {
   /// [package] refers to the package parameter of the intent, can be null.
   /// [componentName] refers to the component name of the intent, can be null.
   /// If not null, then [package] but also be provided.
+  /// [ignoredPackages] is the list of package names that should not by displayed in the selection dialog and
+  /// should not be used to resolve this intent.
```
I think it would also be worth saying this is only used in the `showChooser` method.

NikulinEB replied:
The `ignoredPackages` argument is used in the `launch` method, if the package is not set.

mklim commented on packages/android_intent/lib/android_intent.dart:
```diff
@@ -100,6 +104,10 @@ class AndroidIntent {
   /// See https://developer.android.com/reference/android/content/Intent.html#setComponent(android.content.ComponentName).
   final String componentName;
+
+  /// List of package names that will be excluded from the list of applications to resolve the intent.
+  final List<String> ignoredPackages;
```
Same here, I think it's worth saying this is only being used for `showChooser`.

NikulinEB commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -59,6 +63,26 @@ void send(Intent intent) {
+    Intent chooser = null;
+    intent.setPackage(null);
+    intent.setComponent(null);
+    Boolean isPackageFilterRequired = ignoredPackages != null && ignoredPackages.size() > 0;
```
Thanks for the comments. I added `ignoredPackages` initialization in the `onMethodCall` method (MethodCallHandlerImpl.java). Tell me if this is not what you meant.

NikulinEB commented:
Hi @matthew-carroll 
Could you please review the updated pull request and approve it, if everything is fine?
And tell me, please, should I do something with the check that was not successful?

matthew-carroll commented on packages/android_intent/android/src/test/java/io/flutter/plugins/androidintent/MethodCallHandlerImplTest.java:
```diff
@@ -283,4 +285,38 @@ public void onMethodCall_setsDataAndType() {
   }
+
+  @Test
+  public void onMethodCall_setsIgnoredPackages() {
```
Can you describe what this is testing? As far as I can tell, this test is verifying that including ignored packages doesn't impact the result at all. But isn't it also possible that sometimes the ignored packages will impact the behavior? Shouldn't there a best for that, too?

NikulinEB replied:
I would say that it tests that the package that was added to the `ignoredPackages` is not used to resolve the intent.
I will try to update the test to explicitly verify that the specified package to ignore was excluded.

matthew-carroll commented on packages/android_intent/android/src/test/java/io/flutter/plugins/androidintent/MethodCallHandlerImplTest.java:
```diff
@@ -283,4 +285,38 @@ public void onMethodCall_setsDataAndType() {
+  }
+
+  @Test
+  public void onMethodCall_showChooser() {
```
I'm not sure what this test is verifying, either. Shouldn't there be some verification that the new "show chooser" behavior did something? Maybe I'm overlooking some details in this test, but it looks like it's verifying the same kind of behavior that we'd expect if we ran the regular `Intent` behavior, right?

NikulinEB replied:
You're right, I will fix this test.

matthew-carroll commented on packages/android_intent/lib/android_intent.dart:
```diff
@@ -27,6 +27,8 @@ class AndroidIntent {
   /// [package] refers to the package parameter of the intent, can be null.
   /// [componentName] refers to the component name of the intent, can be null.
   /// If not null, then [package] but also be provided.
+  /// [ignoredPackages] is the list of package names that should not by displayed in the selection dialog and
```
type: "by" -> "be"

matthew-carroll commented on packages/android_intent/lib/android_intent.dart:
```diff
@@ -152,6 +160,26 @@ class AndroidIntent {
+  /// [package] and [componentName] will be ignored.
+  ///
+  /// This works only on Android platforms.
+  Future<void> showChooser() async {
```
I'll connect my earlier question to this point here. When a developer invokes this method, does that developer expect that a chooser will definitely be displayed? If so, how can the developer take action in the case that no `Activity` is available, or in the case that no target `Activity`s exist?

matthew-carroll commented on packages/android_intent/test/android_intent_test.dart:
```diff
@@ -140,6 +140,52 @@ void main() {
+        }
+      });
+
+      test('pass right params', () async {
```
I don't think `"pass right params"` is a name that future developers will be able to maintain. Can you please specify what condition you're testing so that there's no confusion in the future? In other words, a developer might wonder "what are 'right' params and what are 'wrong' params?" and "is there only 1 possible set of 'right' params?"

matthew-carroll commented on packages/android_intent/test/android_intent_test.dart:
```diff
@@ -140,6 +140,52 @@ void main() {
+        }));
+      });
+
+      test('call in ios platform', () async {
```
Rather than naming a test by what it does, consider naming the test based on what you're verifying. For example, this test might be called `'does nothing when called on iOS'`.

NikulinEB replied:
Just used the same name as in the existing test. No problem, I will fix it.

matthew-carroll commented on packages/android_intent/test/android_intent_test.dart:
```diff
@@ -140,6 +140,52 @@ void main() {
       });
     });
+
+    group('showChooser', () {
```
Related to a couple of my earlier questions, I think that tests should probably exist in this group based on a situation where no target activity is found, and any other situation that would result in a chooser not being shown.

matthew-carroll commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -59,6 +64,30 @@ void send(Intent intent) {
+    intent.setComponent(null);
+
+    // Exclude ignored packages from the displayed list.
+    if (ignoredPackages.size() > 0) {
```
Is this if-statement correct? Why do we need to ignore packages to show a chooser?

NikulinEB replied:
This is a mistake, thanks

NikulinEB commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -59,6 +64,30 @@ void send(Intent intent) {
+      }
+    }
+
+    if (chooser != null) {
```
The reason for having this method is not for some specific purpose for using it. This method appeared to simplify the `buildIntent` method by removing the corresponding `showChooser` parameter from it. This parameter appeared to save the previous method behavior. Earlier, when calling the `send` method of this class, the default chooser might appear or not, depending on whether the default application is set to resolve this intent. That is, the appearance of the chooser was not controlled in any way by the developer (unless explicitly specifying a package so as not to show the chooser).

When I implemented ignoring the specified packages, there were two options: use any available activity to resolve this intent (using the `intent.setPackage` method), if it is not in the ignore list - in this case, the chooser will never appear because the activity is already selected. Another option was to show the chooser with all available activities except ignored. Therefore, this `showChooser` method appeared, so that the developer can explicitly choose this method to call, if he prefers to show the chooser.

If the chooser is null, then no available activities were found and the same thing will happen as if we called the `send` method for this intent.

What do you think about the need for this method or its name?

NikulinEB commented on packages/android_intent/android/src/main/java/io/flutter/plugins/androidintent/IntentSender.java:
```diff
@@ -158,7 +189,46 @@ Intent buildIntent(
+  private List<ResolveInfo> getIntentActivities(Intent intent) {
+    PackageManager packageManager;
+    if (activity != null) {
+      packageManager = activity.getPackageManager();
```
I'm not sure, but it seems to me that the same instance is being returned

NikulinEB commented:
Hello @matthew-carroll! I apologize for the long absence. I updated the tests for displaying a chooser and ignoring packages, and fixed some of the problems you wrote about. Also I added comments, so I hope that now the action of the `showChooser` method is clear.
Please review again, I hope in the last time. Thank you in advance!

NikulinEB commented:
@matthew-carroll Hi! I am here again. Let's finish this pull request. Please review and approve if everything is ok. And who should merge these changes?

matthew-carroll commented:
Hi @NikulinEB, I'm swamped with other things at the moment. Perhaps a member of the team can help get this merged. CC @amirh

stuartmorgan commented:
We apologize for the long delay in following up on this PR. We’ve made the decision to no longer accept non-critical PRs for this plugin, as we hope that in time we’re able to transition users to the corresponding plugin in [the Flutter Community Plus Plugins repository](https://plus.fluttercommunity.dev/), so we encourage you to submit your PR there. I realize it's not ideal that this decision is happening after there has already gone through some review here, but hopefully that feedback will still be useful for a Flutter Community Plus Plugin version of the same change.

We’re in the process of overhauling our PR triage system to respond much more quickly to avoid this kind of delay going forward. (Transitioning plugins like this one to the Flutter Community group is in fact part of that effort.)
